core/field_checkbox.js

```diff
@@ -92,49 +93,72 @@ Blockly.FieldCheckbox.prototype.init = function() {
       this.fieldGroup_);
   var textNode = document.createTextNode(Blockly.FieldCheckbox.CHECK_CHAR);
   this.checkElement_.appendChild(textNode);
-  this.checkElement_.style.display = this.state_ ? 'block' : 'none';
+  this.checkElement_.style.display = this.value_ ? 'block' : 'none';
+
+  if (this.borderRect_) {
+    this.borderRect_.setAttribute('width',
+        this.size_.width + Blockly.BlockSvg.SEP_SPACE_X);
+  }
 };
 
 /**
- * Return 'TRUE' if the checkbox is checked, 'FALSE' otherwise.
- * @return {string} Current state.
+ * Checkboxes have a constant width.
+ * @private
  */
-Blockly.FieldCheckbox.prototype.getValue = function() {
-  return String(this.state_).toUpperCase();
+Blockly.FieldCheckbox.prototype.render_ = function() {
+  this.size_.width = 5;
+};
+
+/**
+ * Toggle the state of the checkbox on click.
+ * @protected
+ */
+Blockly.FieldCheckbox.prototype.showEditor_ = function() {
+  this.setValue(!this.value_);
+};
+
+/**
+ * Validate the input value to a boolean.
+ * @param {*} newValue The input value.
+ * @return {string} The value converted to a 'TRUE' or 'FALSE' value.
+ * @protected
+ */
+Blockly.FieldCheckbox.prototype.doClassValidation_ = function(newValue) {
+  return this.convertValueToBool_(newValue) ? 'TRUE' : 'FALSE';
 };
 
 /**
- * Set the checkbox to be checked if newBool is 'TRUE' or true,
- * unchecks otherwise.
- * @param {string|boolean} newBool New state.
+ * Store the new value of the field and update the checkElement.
+ * @param {boolean|string} newValue The new boolean value of the field.
+ * @protected
  */
-Blockly.FieldCheckbox.prototype.setValue = function(newBool) {
-  var newState = (typeof newBool == 'string') ?
-      (newBool.toUpperCase() == 'TRUE') : !!newBool;
-  if (this.state_ !== newState) {
-    if (this.sourceBlock_ && Blockly.Events.isEnabled()) {
-      Blockly.Events.fire(new Blockly.Events.BlockChange(
-          this.sourceBlock_, 'field', this.name, this.state_, newState));
-    }
-    this.state_ = newState;
-    if (this.checkElement_) {
-      this.checkElement_.style.display = newState ? 'block' : 'none';
-    }
+Blockly.FieldCheckbox.prototype.doValueUpdate_ = function(newValue) {
+  this.value_ = this.convertValueToBool_(newValue);
+  // Update visual.
+  if (this.checkElement_) {
+    this.checkElement_.style.display = this.value_ ? 'block' : 'none';
   }
 };
 
 /**
- * Toggle the state of the checkbox.
+ * Get the value of this field, either 'TRUE' or 'FALSE'.
+ * @return {string} The value of this field.
+ */
+Blockly.FieldCheckbox.prototype.getValue = function() {
```

Comment:
While you're here, add a `getValueBoolean` method on `FieldCheckbox` that just returns the actual value, so that developers don't *have* to go back and forth checking string equality.

Reply:
Fixed